[vcpkg registries] Registries: Take 2

strega-nil opened the pull request:
This RFC is the new state of the world with regard to our plan for registries; it is very different from the [old RFC]. It includes registry databases, which are how one describes the set of versions of ports that one has access to. It also includes the specification for git registries, and updates the specification for filesystem registries with these new registry databases.

Since this RFC completely changes how registries work, the existing implementation will be broken.

[Rendered](https://github.com/strega-nil/vcpkg/blob/git-registries/docs/specifications/registries-2.md)

[old RFC]: https://github.com/microsoft/vcpkg/blob/master/docs/specifications/registries.md

BillyONeal commented on docs/specifications/git-registries.md:
```diff
@@ -0,0 +1,123 @@
+### `vcpkg update`
+
+You'll notice that once the repository is added the first time,
+there is only one way to update the repository to the tag at a later date - deleting the lock file.
```
Can they change the tag though?

ras0219-msft replied:
I think the idea is that the tag will not be re-queried if the lock file exists.

BillyONeal replied:
I mean, if the user edits their vcpkg.json to point to a different tag, do we notice and take action?

strega-nil replied:
Sure, in the `vcpkg-configuration.json` file.

strega-nil replied:
Yes, since `vcpkg-lock` maps `{repository, ref}` pairs to specific commits.

BillyONeal reviewed (approved):
Thanks for your great work!

ras0219-msft commented on docs/specifications/git-registries.md:
```diff
@@ -0,0 +1,123 @@
+Then, for every repository, we'll add a directory named with the sha256 of the repository.
+
+In this repository root, we `git init`, run `git fetch <repository url> <ref>`,
+and then check that commit out at `<repository root>/<commit>`.
```
```suggestion
and then check that commit out at `registries/<commit>`.
```
Do we not want to share commits between registries (as long as the commit hash matches)?

strega-nil replied:
hmm, that works. It also fixes the "different URLs map to the same repo"

strega-nil commented on docs/specifications/git-registries.md:
```diff
@@ -0,0 +1,123 @@
+we will add a new directory to cache the downloaded git registries: `registries/git`.
+Then, for every repository, we'll add a directory named with the sha256 of the repository.
+
+In this repository root, we `git init`, run `git fetch <repository url> <ref>`,
```
I don't know; is there a good way to do this? we might have to just have a lock file.

strega-nil commented on docs/specifications/git-registries.md:
```diff
@@ -0,0 +1,123 @@
+Therefore, we've decided on the following implementation strategy:
+
+In the current vcpkg cache home (shared by binary caching archives,
+located at `${XDG_CACHE_HOME-$HOME/.cache}/vcpkg` on unix, and `%LOCALAPPDATA%\vcpkg` on Windows),
```
There are no environment docs for this right now.

vicroms commented on docs/specifications/registries-2.md:
```diff
@@ -0,0 +1,531 @@
+      "version-string": "1.12.7",
+      "git-tree": "466e96fd2e17dd2453aa31dc0bc61bdcf53e7f61",
+      "dependencies": [
+        { "name": "freeimage", "version-string": "3.18.0", "port-version": 13 },
```
Shouldn't these be version requirements instead of version declarations? 

Since we are adding version information here as a way to impose version requirements on packages that have existed before this was possible.

```json
[{ "name": "freeimage", "version-string": "3.18.0", "port-version": 13 }, ...]
```

strega-nil replied:
You're right.

BillyONeal commented on docs/specifications/registries-2.md:
```diff
@@ -0,0 +1,598 @@
+We don't wish to have the same problem as we do right now,
+where there are nearly 1500 entries in a single directory.
+We solve this by placing each database entry into `😇/<first character of port name>-/<port name>.json`.
+For example, the database entry for 7zip shall be at `😇/7-/7zip.json`.
```
```suggestion
For example, the database entry for 7zip is `😇/7-/7zip.json`.
```

BillyONeal commented on docs/specifications/registries-2.md:
```diff
@@ -0,0 +1,598 @@
+There are two actual implementations on disk to consider here: the implementation of the registry database,
+and once we have the database entries for the ports, accessing the port data from the git tree object.
+
+Both of these implementations shall be placed in the vcpkg cache home (shared by binary caching archives).
```
```suggestion
Both of these implementations are placed in the vcpkg cache home (shared by binary caching archives).
```

BillyONeal commented on docs/specifications/registries-2.md:
```diff
@@ -0,0 +1,598 @@
+### Registry Database
+
+At `registries/git`,
+we'll create a new git repository root which shall contain all information from all git registries,
```
```suggestion
we'll create a new git repository root which contains all information from all git registries,
```

BillyONeal commented on docs/specifications/registries-2.md:
```diff
@@ -0,0 +1,598 @@
+
+## Git Registries: Implementation on Disk
+
+There are two actual implementations on disk to consider here: the implementation of the registry database,
```
```suggestion
There are two implementations on disk to consider here: the implementation of the registry database,
```

BillyONeal commented on docs/specifications/registries-2.md:
```diff
@@ -0,0 +1,598 @@
+
+### Registry Database
+
+At `registries/git`,
```
```suggestion
At `$CACHE_ROOT/registries/git`,
```

?

BillyONeal replied:
(And similar for all paths that need to be relative to that; since it's confusing whether we are talking about the cache location vs. a vcpkg tree from there)

BillyONeal commented on docs/specifications/registries-2.md:
```diff
@@ -0,0 +1,598 @@
+file from the right commit: `git show <commit id> -- 😇/<first character>-/<portname>.json`.
+
+One unfortunate thing about having one directory being used for all vcpkg instances on a machine is
+that it's very possible to have an issue with concurrency - for example, after `fetch`ing the latest HEAD
```
```suggestion
that it's possible to have an issue with concurrency - for example, after `fetch`ing the latest HEAD
```

strega-nil commented on docs/specifications/registries-2.md:
```diff
@@ -0,0 +1,598 @@
+
+The folder structure is up for bikeshedding; however right now the design of it is as follows.
+We don't wish to have the same problem as we do right now,
+where there are nearly 1500 entries in a single directory.
```
Yes absolutely. One advantage of this new registry structure means we can modify the ports tree without breaking anything (assuming people update to the new version of vcpkg).

klalumiere commented:
The RFC is super interesting! :nerd_face: 

Something it doesn't talk about is triplets. Will it be possible to store them in the registry alongside the ports?

A little more details about the context/motivations of my question might help. Right now, in our prototype integration of `vcpkg`, we do something like `vcpkg install mypackage:x64-linux-dynamic --overlay-ports ../vcpkg-overlays/ports --overlay-triplets ../vcpkg-overlays/triplets`. When available, registries and `vcpkg-configuration.json` will "manage" the `--overlay-ports ../vcpkg-overlays/ports` for us. Actually it will do much more: pulling the repository for us, locking the SHA, etc. Will we be able to get rid of the `--overlay-triplets ../vcpkg-overlays/tripplets` part too?

strega-nil commented:
@klalumiere we may consider adding triplet registries at some point, but the plan at this point is to just add overlay triplets to the configuration file.

vicroms commented on docs/specifications/registries-2.md:
```diff
@@ -105,48 +130,18 @@ This example is of `ogre`, since this port has both features and dependencies;
-    ...
-  ]
-}
+[
+  {
+    "version-string": "1.12.7",
+    "git-tree": "466e96fd2e17dd2453aa31dc0bc61bdcf53e7f61",
+  },
+  {
+    "version-string": "1.12.1",
+    "port-version": 1,
+    "git-tree": "0de81b4f7e0ec24966e929c2ea64e16c15e71d5e",
+  },
+  ...
+]
```
Database generator implemented according to spec in PR #13777

strega-nil commented:
I'm calling for an FCP. Please post all comments this next week; assuming no new blocking issues, this specification will be merged at 17:00 PST (UTC-8) 2020-11-06.

Neumann-A commented on docs/specifications/registries-2.md:
```diff
@@ -0,0 +1,634 @@
+[
+  {
+    "version-string": "2020-07-22",
+    "registry-path": "/ports/a-/asmjit/2020-07-22"
```
This looks like magic without knowing the implementation in detail.
Something like `${base}/ports/a-/asmjit/2020-07-22` is in my eyes clearer.
(or I am just VS Code trained where some variables are also available in settings.json)

strega-nil commented:
I agree with @Neumann-A's assessment that the `registry-path: "/foo/bar"` is a bit of unfortunate magic. However, after discussing this with the rest of the team, we think that there are two real options for fixing this:

1. We could use @Neumann-A's suggestion outright; this is pretty simple, but it would require figuring out what we should do with paths that are "weird", for example, `"${base}foo`, `${base}/../foo`, or paths that aren't rooted at `${base}`. As a part of this, `base` might also not be the best option.
2. @BillyONeal suggested "`$` as root", i.e., `$/foo`, and requiring paths to be rooted there. This fixes all of the issues with "weird paths" from 1, since we'd just disallow all of those kinds of paths.

I'd like to take input from the community. If you think that option 1 is better, please respond to this message with 🚀 ; if you think that option 2 is better, please respond to this message with 🎉 . If you have more complex thoughts, please respond in a reply :)

klalumiere commented on docs/specifications/registries-2.md:
````diff
@@ -0,0 +1,634 @@
+We should first copy the existing implementation to a new folder:
+
+```sh
+$ cp -r ports/a-/asmjit/2020-07-22 ports/a-/asmjit/2020-10-08
````
Do you think it would be a good idea to add an option to the `vcpkg add-port` command so that this part is done "automatically" for us? Something like

1. `vcpkg add-port ./ports/a-/asmjit/2020-10-08 --base ports/a-/asmjit/2020-07-22 --port-type registry-path`
1. Make the edits required to `ports/a-/asmjit/2020-10-08` to update to latest.

This would remove a manual manipulation, and it prevents repeating `ports/a-/asmjit/2020-10-08` twice, removing at the same time an opportunity to make a typo. :slightly_smiling_face:

strega-nil committed:
[vcpkg registries] Registries: Take 2

unfortunately without all the history because I really screwed up the git commands

This RFC is the new state of the world with regard to our plan for registries; it is very different from the old RFC. It includes registry databases, which are how one describes the set of versions of ports that one has access to. It also includes the specification for git registries, and updates the specification for filesystem registries with these new registry databases.

Since this RFC completely changes how registries work, the existing implementation will be broken.

vicroms commented on docs/specifications/registries-2.md:
````diff
@@ -0,0 +1,631 @@
+```json
+{
+  "default": {
+    "abseil": { "version-string": "2020-03-03" },
+    "zlib": { "version-string": "1.2.11", "port-version": 9 }
````
I think we should make this a generic version field that does not imply any kind of scheme. 

I would suggest `"version"` but it is kind of unfortunate that it overlaps with the "Relaxed" scheme.
```
"abseil": { "version": "2020-03-03" },
"zlib": { "version": "1.2.11", "port-version": 9 }
```

Another suggestion would be "version-text"
```
"abseil": { "version-text": "2020-03-03" }
"zlib": { "version-text": "1.2.11", "port-version": 9 }
```
Or we could change the spec for versioning so that `"version"` does not imply relaxed and instead make it explicit `"version-relaxed"`.

vicroms commented on docs/specifications/registries-2.md:
````diff
@@ -0,0 +1,631 @@
+```json
+[
+  {
+    "version-string": "1.12.7",
+    "git-tree": "466e96fd2e17dd2453aa31dc0bc61bdcf53e7f61",
````
Same as with baselines use a generic version field.

```
{
  "version": "1.12.7",
  "git-tree": "466e96fd2e17dd2453aa31dc0bc61bdcf53e7f61"
}
```

strega-nil replied:
This definitely shouldn't use a generic field; there's no other place to grab the version scheme from.

strega-nil committed:
update the registries RFC for modern design

strega-nil committed:
remove bikeshedding

BillyONeal commented:
@strega-nil Given that this feature has now landed, shouldn't we merge this RFC?

strega-nil commented:
@BillyONeal yes.
